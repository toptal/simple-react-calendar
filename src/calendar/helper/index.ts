import addDays from 'date-fns/add_days'
import addMonths from 'date-fns/add_months'
import addWeeks from 'date-fns/add_weeks'
import differenceInWeeks from 'date-fns/difference_in_weeks'
import eachDay from 'date-fns/each_day'
import endOfMonth from 'date-fns/end_of_month'
import endOfWeek from 'date-fns/end_of_week'
import format from 'date-fns/format'
import isAfter from 'date-fns/is_after'
import isBefore from 'date-fns/is_before'
import isEqual from 'date-fns/is_equal'
import isSameDay from 'date-fns/is_same_day'
import isSameMonth from 'date-fns/is_same_month'
import isValidDate from 'date-fns/is_valid'
import isWeekend from 'date-fns/is_weekend'
import isWithinRange from 'date-fns/is_within_range'
import parse from 'date-fns/parse'
import startOfDay from 'date-fns/start_of_day'
import startOfMonth from 'date-fns/start_of_month'
import startOfWeek from 'date-fns/start_of_week'
import startOfYear from 'date-fns/start_of_year'
import subDays from 'date-fns/sub_days'

import { IDate, IDateRange, ISelection, ISelectionRange } from '../../@types'

export const EMPTY_RANGE = {start: undefined, end: undefined}

interface IGetActiveMonth {
  activeMonth?: IDate
  mode: 'range' | 'single'
  today: Date
  selected?: IDate | ISelectionRange
}

export const getActiveMonth = ({selected, activeMonth, mode, today}: IGetActiveMonth): Date => {
  if (!activeMonth) {
    return startOfMonth(today)
  } else {
    if (isValid(activeMonth as Date)) {
      return startOfMonth(activeMonth as IDate)
    }

    const selectionStart = mode === 'single' ? selected : (selected as ISelectionRange).start

    if (!isValid(selectionStart as Date)) {
      console.warn('SimpleReactCalendar, getActiveMonth, mode is range but selection start is not a valid Date')
    }

    return startOfMonth((selectionStart as Date) || today)
  }
}

export const getAsDate = (day: IDate) => {
  const parsedDate = parse(day)

  return isValidDate(parsedDate) ? parsedDate : new Date()
}

interface IGetMonthIsEnabled {
  activeMonth: IDate
  maxDate?: IDate
  minDate?: IDate
  type: 'prev' | 'next'
}

interface IGetDay {
  day: IDate
  offsetInDays: number
}

export const getDay = ({day, offsetInDays}: IGetDay) => addDays(day, offsetInDays || 0)

interface IGetDayIsBefore {
  date: IDate
  dateToCompare: IDate
}

export const getDayIsBefore = ({date, dateToCompare}: IGetDayIsBefore) => isBefore(date, dateToCompare)

export const getIsRangeOverlap = () => {}

export const getMonthIsEnabled = ({activeMonth, maxDate, minDate, type}: IGetMonthIsEnabled): boolean =>
  type === 'next'
    ? isAfter(startOfMonth(maxDate as IDate), startOfMonth(activeMonth))
    : isBefore(startOfMonth(minDate as IDate), startOfMonth(activeMonth))

export interface IGetDaysOfWeek {
  daysOfWeek: string[]
  dayIndex: 0 | 1 | 2 | 3 | 4 | 5 | 6
}

export const getDaysOfWeek = ({daysOfWeek, dayIndex}: IGetDaysOfWeek): string[] => {
  const adjustedIndex = dayIndex - 1

  return daysOfWeek.slice(adjustedIndex).concat(daysOfWeek.slice(0, adjustedIndex))
}

export const isValid = (date: Date) => {
  try {
    return isValidDate(date)
  } catch (e) {
    return false
  }
}

interface IGetHighlighted {
  highlighted?: {start: Date; end: Date}
}

export const getHighlighted = ({highlighted}: IGetHighlighted) => {
  if (!highlighted) return EMPTY_RANGE

  const {start, end} = highlighted

  if (isValid(start as Date) && isValid(end as Date)) {
    return highlighted
  } else {
    return EMPTY_RANGE
  }
}

interface IGetSelectionDate extends IGetSelection {
  dateType: 'start' | 'end'
}

export const getSelectionDate = ({
  selected,
  onSelectionProgress,
  stateSelection,
  mode,
  dateType,
}: IGetSelectionDate) => {
  switch (mode) {
    case 'single':
      return selected

    case 'range':
      if (!onSelectionProgress && stateSelection) {
        return stateSelection[dateType]
      } else {
        return selected && (selected as ISelectionRange)[dateType]
      }
  }
}

interface IGetSelection {
  selected?: IDate | ISelectionRange
  onSelectionProgress?: (changes: ISelection) => void
  stateSelection?: ISelection
  mode: 'range' | 'single'
}

export const getSelection = ({selected, onSelectionProgress, stateSelection, mode}: IGetSelection) => {
  const start = getSelectionDate({selected, onSelectionProgress, stateSelection, mode, dateType: 'start'})

  const end = getSelectionDate({selected, onSelectionProgress, stateSelection, mode, dateType: 'end'})

  if (isValid(start) && isValid(end)) {
    return {start, end}
  } else {
    return EMPTY_RANGE
  }
}

export interface IGetMonthWeeks {
  activeMonth: Date
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6
  minNumberOfWeeks?: number
}

export interface IMonthWeeks {
  weekStart: Date
  weekEnd: Date
  weekNumber: number
}

export const getMonthWeeks = ({activeMonth, minNumberOfWeeks, weekStartsOn}: IGetMonthWeeks): IMonthWeeks[] => {
  const startOfTheMonth = startOfWeek(startOfMonth(activeMonth), {weekStartsOn})
  const endOfTheMonth = endOfWeek(endOfMonth(activeMonth), {weekStartsOn})
  const startOfTheYear = startOfYear(activeMonth)
  const lengthNumber = minNumberOfWeeks || differenceInWeeks(endOfTheMonth, startOfTheMonth) + 1

  return Array.from(new Array(lengthNumber)).map((_, index) => {
    const weekStart = addWeeks(startOfTheMonth, index)

    return {
      weekStart,
      weekNumber: differenceInWeeks(weekStart, startOfTheYear) + 2,
      weekEnd: addDays(weekStart, 7),
    }
  })
}

interface IGetMaxDate extends IMinMaxDate {
  maxDate?: IDate
}

export const getMaxDate = ({maxDate, rangeLimit, selectionStart}: IGetMaxDate): IDate => {
  const calcEndDate = addDays(selectionStart, rangeLimit || 0)

  if (maxDate) {
    return isBefore(calcEndDate, maxDate) ? calcEndDate : maxDate
  } else {
    return calcEndDate
  }
}

interface IMinMaxDate {
  rangeLimit?: number
  selectionStart: Date
}

interface IGetMinDate extends IMinMaxDate {
  minDate?: IDate
}
export const getMinDate = ({rangeLimit, minDate, selectionStart}: IGetMinDate): IDate => {
  const calcStartDate = subDays(selectionStart, rangeLimit || 0)

  if (minDate) {
    return isBefore(minDate, calcStartDate) ? calcStartDate : minDate
  } else {
    return calcStartDate
  }
}

export interface IGetDayIsWeekend {
  day: Date
  // weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6
}

export const getDayIsWeekend = ({day}: IGetDayIsWeekend) => {
  // TODO: Should respects the weekStartOn

  return isWeekend(day)
}

export interface IGetDayIsSame {
  date: IDate
  dateToCompare: IDate
}

export const getDayIsSame = ({date, dateToCompare}: IGetDayIsSame) => isSameDay(date, dateToCompare)

export interface IGetDayIsSelection {
  date: Date
  type?: 'start' | 'end'
  selection: IDate | ISelectionRange
}

export const getDayIsSelection = ({type, date, selection}: IGetDayIsSelection) => {
  const isRange = !isValid(selection as Date)

  if (!isRange) {
    return getDayIsSame({date1: selection as Date, date2: date})
  }

  const {start, end} = selection as ISelectionRange

  if (!type) {
    return isWithinRange(day, start, end)
  } else {
    return getDayIsSame({date1: type === 'start' ? (start as Date) : (end as Date), date2: day})
  }
}

export interface IGetDayIsDisabled {
  date: Date
  disabledIntervals?: IDateRange[]
}

export const getDayIsDisabled = ({date, disabledIntervals}: IGetDayIsDisabled) => {
  if (!disabledIntervals || !disabledIntervals.length) return false

  return disabledIntervals
    .map(({start, end}) => isWithinRange(startOfDay(date), startOfDay(start), startOfDay(end)))
    .includes(true)
}

export const getRangeIncludesDisabledInterval = ({start, end, disabledIntervals}) => {}

export interface IGetDayIsNextMonth {
  date: IDate
  activeMonth: IDate
}

export const getDayIsNextMonth = ({date, activeMonth}: IGetDayIsNextMonth) =>
  !isSameMonth(date, activeMonth) && isAfter(date, activeMonth)

export const getDayISOFormatter = (date: Date) => format(date, 'YYYY-MM-DD')

export const getMonthFormattedName = (date: Date) => format(date, 'MMMM YYYY')

interface IGetDayIsHighlighted {
  date: IDate
  startDay: IDate
  endDay: IDate
}

export const getDayIsHighlighted = ({date, startDay, endDay}: IGetDayIsHighlighted) =>
  isWithinRange(startOfDay(date), startOfDay(startDay), startOfDay(endDay))

interface IGetDayIsCurrentMonth {
  activeMonth: IDate
  date: IDate
}

export const getDayIsCurrentMonth = ({activeMonth, date}: IGetDayIsCurrentMonth) => isSameMonth(date, activeMonth)

interface IGetDayIsSelectable {
  date: IDate
  minDate: IDate
  maxDate: IDate
  disabledIntervals: IDateRange[]
}

export const getDayIsSelectable = ({date, minDate, maxDate, disabledIntervals}: IGetDayIsSelectable) => {
  if (getDayIsDisabled({date, disabledIntervals})) return false

  if (minDate && maxDate) {
    return isWithinRange(date, minDate, maxDate)
  } else if (minDate && !maxDate) {
    return isAfter(date, minDate) || isEqual(date, minDate)
  } else if (maxDate && !minDate) {
    return isBefore(date, maxDate) || isEqual(date, maxDate)
  } else {
    return true
  }
}

export const getDayEachDay = ({start, end}: IDateRange) => eachDay(start, end)

interface IGetMonth {
  date: IDate
  offsetInMonth: number
}

export const getMonth = ({date, offsetInMonth}: IGetMonth) => addMonths(date, offsetInMonth)

export const getIntervalsAsDate = (intervals?: IDateRange[]) => {
  if (!intervals || !intervals.length) return

  return intervals.map(({start, end}) => ({start: getAsDate(start), end: getAsDate(end)}))
}

export const getSelectedAsDate = (selected?: IDate | ISelectionRange) => {
  if (!selected) return

  if (!(selected as ISelectionRange).start) {
    return getAsDate(selected as IDate)
  } else {
    return {
      inProgress: (selected as ISelectionRange).inProgress || false,
      start: getAsDate((selected as ISelectionRange).start),
      end: getAsDate((selected as ISelectionRange).end),
    }
  }
}

export const getDayFormatted = (date: Date) => format(date, 'd')
