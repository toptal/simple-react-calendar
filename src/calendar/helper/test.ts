import { getMonthWeeks } from '.'

describe('Helper functions', () => {
  describe.only('#getMonthWeeks', () => {
    describe('when no `minNumberOfWeeks` has been passed', () => {
      test('returns a weeks array', () => {
        expect(getMonthWeeks({ activeMonth: new Date(2019, 10, 1), weekStartsOn: 1 })).toEqual([
          { weekStart: '2019-10-28', weekNumber: 44 },
          { weekStart: '2019-11-04', weekNumber: 45 },
          { weekStart: '2019-11-11', weekNumber: 46 },
          { weekStart: '2019-11-18', weekNumber: 47 },
          { weekStart: '2019-11-25', weekNumber: 48 },
        ])
      })
    })

    describe('when `minNumberOfWeeks` has been passed', () => {
      test('returns a weeks array', () => {
        expect(getMonthWeeks({ activeMonth: new Date(2019, 10, 1), minNumberOfWeeks: 8, weekStartsOn: 1 })).toEqual([
          { weekStart: '2019-10-28', weekNumber: 44 },
          { weekStart: '2019-11-04', weekNumber: 45 },
          { weekStart: '2019-11-11', weekNumber: 46 },
          { weekStart: '2019-11-18', weekNumber: 47 },
          { weekStart: '2019-11-25', weekNumber: 48 },
          { weekStart: '2019-12-02', weekNumber: 49 },
          { weekStart: '2019-12-09', weekNumber: 50 },
          { weekStart: '2019-12-16', weekNumber: 51 },
        ])
      })
    })
  })
})
