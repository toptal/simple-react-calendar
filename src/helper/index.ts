import formatDate from 'date-fns/format'

export type NoticeMessageType =
  | 'overlapping_with_disabled'
  | 'disabled_day_click'
  | null

export const getISODate = (date: Date) => formatDate(date, 'YYYY-MM-DD')
export const getDayFormatted = (date: Date) => formatDate(date, 'D')
export const getNoticeContent = (type: NoticeMessageType) => {
  switch (type) {
    case 'overlapping_with_disabled':
    case 'disabled_day_click':
      return 'Selected range must not overlap with disabled dates.'

    default:
      return null
  }
}
