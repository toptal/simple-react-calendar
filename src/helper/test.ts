import { getDayFormatted, getISODate } from '.'

describe('Simple React Calendar - Date helpers', () => {
  describe('#getISODate', () => {
    expect(getISODate(new Date(2020, 1, 9))).toBe('2020-02-09')
  })

  describe('#getDayFormatted', () => {
    expect(getDayFormatted(new Date(2020, 1, 9))).toBe('9')
  })
})
