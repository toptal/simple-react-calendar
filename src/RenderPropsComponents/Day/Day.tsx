import React, { FC } from 'react'
import cx from 'classnames'

import * as helper from '../../helper'
import { HandleOnDayClick, HandleOnDayEnter } from '../../@types'

interface Props {
  blockClassName: string
  date: Date
  getDayFormatted: typeof helper.getDayFormatted
  handleOnClick: HandleOnDayClick
  handleOnEnter: HandleOnDayEnter
  isCurrentMonth: boolean
  isDisabled: boolean
  isHighlighted: boolean
  isMonthNext: boolean
  isMonthPrev: boolean
  isNonSelectable: boolean
  ISODate: string
  isSelectable: boolean
  isSelected: boolean
  isSelectionEnd: boolean
  isSelectionStart: boolean
  isToday: boolean
  isWeekend: boolean
  isWorkDay: boolean
  key?: string
}

const Day: FC<Props> = ({
  blockClassName,
  date,
  getDayFormatted,
  handleOnClick,
  handleOnEnter,
  isCurrentMonth,
  isDisabled,
  isHighlighted,
  isMonthNext,
  isMonthPrev,
  isNonSelectable,
  ISODate,
  isSelectable,
  isSelected,
  isSelectionEnd,
  isSelectionStart,
  isToday,
  isWeekend,
  isWorkDay
}) => (
  <button
    data-simple-react-calendar-day={ISODate}
    className={cx(`${blockClassName}-week-day`, {
      'is-current_month': isCurrentMonth,
      'is-disabled': isDisabled,
      'is-end_selection': isSelectionEnd,
      'is-highlighted': isHighlighted,
      'is-next_month': isMonthNext,
      'is-not_selectable': isNonSelectable,
      'is-prev_month': isMonthPrev,
      'is-selectable': isSelectable,
      'is-selected': isSelected,
      'is-start_selection': isSelectionStart,
      'is-today': isToday,
      'is-weekend': isWeekend,
      'is-working_day': isWorkDay
    })}
    onClick={handleOnClick}
    onMouseEnter={handleOnEnter}
    type='button'
    value={ISODate}
  >
    {getDayFormatted(date)}
  </button>
)

Day.displayName = 'Day'

export default Day
