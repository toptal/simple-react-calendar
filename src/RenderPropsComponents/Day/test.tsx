import { fireEvent, render } from '@testing-library/react'
import React, { ComponentProps } from 'react'

import { getDayFormatted } from '../../helper'
import Day from '.'

const renderDay = (props: Partial<ComponentProps<typeof Day>>) => {
  const baseProps = {
    ISODate: '2020-02-09',
    blockClassName: 'example-class-name',
    date: new Date(2020, 1, 9),
    getDayFormatted,
    handleOnClick: jest.fn(),
    handleOnEnter: jest.fn(),
    isCurrentMonth: true,
    isDisabled: false,
    isHighlighted: true,
    isMonthNext: false,
    isMonthPrev: false,
    isNonSelectable: true,
    isSelectable: true,
    isSelected: false,
    isSelectionEnd: false,
    isSelectionStart: true,
    isToday: true,
    isWeekend: true,
    isWorkDay: false,
    ...props
  }

  // eslint-disable-next-line react/jsx-props-no-spreading
  return render(<Day {...baseProps} />)
}

test('default render', () => {
  const { container } = renderDay({})

  expect(container.firstChild).toMatchSnapshot()
})

test('trigger callback on click', () => {
  const handleOnClick = jest.fn()
  const { getByRole } = renderDay({ handleOnClick })

  const button = getByRole('button')

  fireEvent.click(button)

  expect(handleOnClick).toBeCalled()
})

test('trigger callback on mouseenter', () => {
  const handleOnEnter = jest.fn()
  const { getByRole } = renderDay({ handleOnEnter })

  const button = getByRole('button')

  fireEvent.mouseEnter(button)

  expect(handleOnEnter).toBeCalled()
})

test('handle currentMonth', () => {
  const { container } = renderDay({ isCurrentMonth: true })

  expect(container.firstChild).toHaveClass('is-current_month')
})

test('handle isDisabled', () => {
  const { container } = renderDay({ isDisabled: true })

  expect(container.firstChild).toHaveClass('is-disabled')
})

test('handle isHighlighted', () => {
  const { container } = renderDay({ isHighlighted: true })

  expect(container.firstChild).toHaveClass('is-highlighted')
})

test('handle isMonthNext', () => {
  const { container } = renderDay({ isMonthNext: true })

  expect(container.firstChild).toHaveClass('is-next_month')
})

test('handle isMonthPrev', () => {
  const { container } = renderDay({ isMonthPrev: true })

  expect(container.firstChild).toHaveClass('is-prev_month')
})

test('handle isNonSelectable', () => {
  const { container } = renderDay({ isNonSelectable: true })

  expect(container.firstChild).toHaveClass('is-not_selectable')
})

test('handle isSelectable', () => {
  const { container } = renderDay({ isSelectable: true })

  expect(container.firstChild).toHaveClass('is-selectable')
})

test('handle isSelected', () => {
  const { container } = renderDay({ isSelected: true })

  expect(container.firstChild).toHaveClass('is-selected')
})

test('handle isSelectionEnd', () => {
  const { container } = renderDay({ isSelectionEnd: true })

  expect(container.firstChild).toHaveClass('is-end_selection')
})

test('handle isSelectionStart', () => {
  const { container } = renderDay({ isSelectionStart: true })

  expect(container.firstChild).toHaveClass('is-start_selection')
})

test('handle isToday', () => {
  const { container } = renderDay({ isToday: true })

  expect(container.firstChild).toHaveClass('is-today')
})

test('handle isWeekend', () => {
  const { container } = renderDay({ isWeekend: true })

  expect(container.firstChild).toHaveClass('is-weekend')
})

test('handle isWorkday', () => {
  const { container } = renderDay({ isWorkDay: true })

  expect(container.firstChild).toHaveClass('is-working_day')
})
