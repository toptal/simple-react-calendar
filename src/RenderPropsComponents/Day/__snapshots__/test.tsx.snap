// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default render 1`] = `
<button
  class="example-class-name-week-day is-current_month is-highlighted is-not_selectable is-selectable is-start_selection is-today is-weekend"
  data-simple-react-calendar-day="2020-02-09"
  type="button"
  value="2020-02-09"
>
  9
</button>
`;
